### PR TITLE
Do not extract object for header sources

### LINKED ISSUE #2716: Build error when using target.extract_all_objects() when target has header sources
An issue has been discovered while working on PR #2711:

> ninja: error: 'gio/gio-2.0@sha/meson-generated_.._gnetworking.h.o', needed by 'gio/libgio-2.0.a', missing and no known rule to make it

According to @nirbheek it is a bug in extract_all_objects():
> I've seen this bug before elsewhere. It's because when figuring out the list of sources to use in extract_all_objects(), we don't skip over headers. See the sources classification in backends/ninjabackend.py:NinjaBackend.generate_target(). Specifically, look at how we handle the contents oftarget_sources and generated_sources.
>
>This file classification needs to be factored out and probably moved to the BuildTarget object so it can also be used by extract_all_objects().

### PR DESCRIPTION
**Pull request by xclaesse**
Closes #2716

### CODE REVIEW

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -342,6 +342,13 @@ def object_filename_from_source(self, target, source, is_unity):
         assert isinstance(source, mesonlib.File)
         build_dir = self.environment.get_build_dir()
         rel_src = source.rel_to_builddir(self.build_to_src)
+
+        if (not self.environment.is_source(rel_src) or \
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

the backslash is redundant between brackets

File: mesonbuild/backend/backends.py [hunk 2]
```diff
@@ -342,6 +342,13 @@ def object_filename_from_source(self, target, source, is_unity):
...
+            self.environment.is_header(rel_src)) and \
+           not self.environment.is_object(rel_src) and \
+           not self.environment.is_library(rel_src)
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

SyntaxError: invalid syntax

File: mesonbuild/backend/backends.py [hunk 3]
```diff
@@ -342,6 +342,13 @@ def object_filename_from_source(self, target, source, is_unity):
...
+            return None
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

unexpected indentation

**Comment by xclaesse** [comment]
I'm not sure at all about that condition. I copied the same logic than in ninjabackend, but I'm wondering if it could be just:

if not is_source() or is_header():
    return None

Because libraries and object files should be ignored as well there, no?

File: mesonbuild/backend/backends.py [hunk 4]
```diff
@@ -342,6 +342,11 @@ def object_filename_from_source(self, target, source, is_unity):
         assert isinstance(source, mesonlib.File)
         build_dir = self.environment.get_build_dir()
         rel_src = source.rel_to_builddir(self.build_to_src)
+
+        if not self.environment.is_source(rel_src) or \
+           self.environment.is_header(rel_src))
```
Comment (ignatenkobrain) [inline mesonbuild/backend/backends.py]: __[Flake8]__

SyntaxError: invalid syntax

**Comment by xclaesse** [comment]
Ok, since that function always append '.o' to the return value, I don't think it makes sense on library or object files. So let's restrict to source files that are not headers.

**Comment by xclaesse** [comment]
It seems CI failed because "Build execution time has reached the maximum allowed time for your plan (60 minutes)."

**Comment by jpakkane** [comment]
Also this which seems relevant:

    Traceback (most recent call last):
      File "/root/mesonbuild/mesonmain.py", line 365, in run
        app.generate()
      File "/root/mesonbuild/mesonmain.py", line 151, in generate
        self._generate(env)
      File "/root/mesonbuild/mesonmain.py", line 208, in _generate
        g.generate(intr)
      File "/root/mesonbuild/backend/ninjabackend.py", line 210, in generate
        self.generate_target(t, outfile)
      File "/root/mesonbuild/backend/ninjabackend.py", line 456, in generate_target
        obj_list += self.flatten_object_list(target)
      File "/root/mesonbuild/backend/backends.py", line 235, in flatten_object_list
        obj_list += self.determine_ext_objs(target, obj, proj_dir_to_build_root)
      File "/root/mesonbuild/backend/backends.py", line 392, in determine_ext_objs
        objname = objname.replace('/', '_').replace('\\', '_')
    AttributeError: 'NoneType' object has no attribute 'replace'

**Comment by xclaesse** [comment]
Hmm, I was not expecting this, why is the unity source does not pass the "is_source() and not is_header()" condition? Aren't them just a huge .c file ?

**Comment by xclaesse** [comment]
I don't know why it's needed, but this should fix the CI then.

**Pull request merged**